### DISCUSSION TITLE
Plugin to preprocess a Python docstring?

### DISCUSSION THREAD

**Discussion by WarrenWeckesser** [mkdocstrings/mkdocstrings#581] [Q&A] [answered]
I'm looking into converting the documentation of a Python project from Sphinx to MkDocs.  In this project I use a simple Sphinx extension written in Python that is given the Python object, its docstring as a list of strings, and some other Sphinx configuration data.  The extension modifies the text of the docstring in place; regexes are used to find patterns and make substitutions.  This extension is called before Sphinx processes the markup in the docstring.  The regex's can be defined in Sphinx's `conf.py` (for "global" replacements), and they can be attached as attributes on the Python object (for object-specific replacements).  The actual code for the Sphinx extension is here: https://github.com/WarrenWeckesser/mpsci/blob/main/doc/sphinxext/resubber.py.

So far I haven't figured out if something like this is possible with mkdocstrings or its Python handler.  Is there anything like this?

I use this extension to replace certain text strings with LaTeX markup.  In Python, docstrings can be viewed "raw" in an interactive shell, and presenting LaTeX directly to users is nasty.  With this plugin, I can write a docstring using human-readable text, and still have nicely rendered math output from Sphinx to be displayed in a web page.

**Comment by pawamoy** [comment]
Hi @WarrenWeckesser, good news, this is possible :slightly_smiling_face: 

Griffe, the library used by the Python handler, supports extensions. I'm currently rewriting its extension system, but you can start using the existing one if you want (don't worry, I'll add a proper deprecation period if needed, and the new system will be very similar, just more flexible and easy to use).

```python
# your_extension.py
import ast

from griffe.dataclasses import Object
from griffe.extensions import VisitorExtension


def update_docstring(obj: Object) -> str:
    # your logic using obj.docstring.value


def update_docstrings_recursively(obj: Object) -> None:
    if obj.docstring:
        obj.docstring.value = update_docstring(obj)
    for member in obj.members.values():
        if not member.is_alias:
            update_docstrings_recursively(member)


class YourExtension(VisitorExtension):
    def visit_module(self, node: ast.AST) -> None:
        module = self.visitor.current.module
        update_docstrings_recursively(module)


Exension = YourExtension
```

```yaml
# mkdocs.yml
plugins:
- mkdocstrings:
    handlers:
      python:
        options:
          extensions:
          - your_extension.py
```

There are other ways to write extensions that follow the loading process more closely, but I don't think that's needed here.
Checkout the API reference for Object and other classes: https://mkdocstrings.github.io/griffe/reference/griffe/dataclasses/#griffe.dataclasses.Object.

Moving this to a discussion by the way :slightly_smiling_face:

**Reply by WarrenWeckesser** [reply]
Thanks @pawamoy, that's great!  However, to reproduce the Sphinx extension, I need the objects that are created when the code is imported.  For example, for a function, I need the actual Python function that is created--and to which additional attributes are added--when the module that it is defined in is imported, because I store the text substitutions to be made in an attribute of the function.  I don't see a reference to the actual function in `griffe.dataclasses.Function`.  Does griffe run an import on the code?  Or is it only parsing the Python source, and not actually executing it?  If the latter, I'll have to come up with a different method for configuring the object-specific substitutions.

**Reply by pawamoy** [reply, marked as answer]
Hey @WarrenWeckesser, sorry, I forgot answering to you.

I very recently reworked the extension system. Check out the full example in the docs: https://mkdocstrings.github.io/griffe/extensions/#full-example. You'll see that I make use of [`dynamic_import`](https://mkdocstrings.github.io/griffe/reference/griffe/importer/#griffe.importer.dynamic_import) to import objects while running a static analysis. I think this is what you wanted!

**Reply by WarrenWeckesser** [reply]
Thanks @pawamoy.  I'll let you know when I've had a chance to give it a try.
